### PR TITLE
add basepath support

### PR DESCRIPTION
**Pull request by pmjones**
This should be a fix for #86 -- @harikt @pine3ree please review and comment.

Basically, you create a RouterFactory with the basepath you want; that would be the subdirectory in which the application lives. You can then add routes as normal, without having to specify the basepath prefix; the matcher will match properly, and the generator will generate paths with the basepath prefix in place.

### CODE REVIEW

File: tests/RouterTest.php [hunk 1]
```diff
@@ -498,4 +498,22 @@ public function testAddWithAction()
...
+
+        // should fail without basepath
+        $this->assertFalse($this->router->match('/foo/bar'));
+
+        // should pass with basepath
+        $actual = $this->router->match('/path/to/sub/index.php/foo/bar');
+        $this->assertSame('DirectAction', $actual->values['action']);
+
+        // should get the basepath in place
+        $expect = '/path/to/sub/index.php/foo/bar';
```
Comment (harikt) [inline tests/RouterTest.php:515]: good to have one more test with no index.php . So is `/path/to/sub/foo/bar`

**Comment by harikt** [comment]
Thank you @pmjones . This looks good to me.

**Comment by pine3ree** [comment]
Thank You @pmjones 
Looks fine to me.
To be on the safe side i would add 

``` php
//...
$this->basepath = rtrim($basepath, ' /');
```

 to the Router constructor for handling params like '/path/to/sub/'

**Comment by pmjones** [comment]
Added a dir-only test, and added rtrim. When the CI passes, I'll merge it. Thanks guys!

**Pull request merged**